orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -244,6 +287,178 @@ Status AddOrSkipRecordForwardWaitBackward(Graph& graph,
   return Status::OK();
 }
 
+// Replace node_args[i] with new_node_args[i] for all inputs in nodes.
+void ReplaceNodeArgs(std::vector<Node*>& nodes,
+                     std::vector<NodeArg*>& node_args,
+                     std::vector<NodeArg*>& new_node_args) {
+  ORT_ENFORCE(node_args.size() == new_node_args.size());
+  for (size_t i = 0; i < node_args.size(); ++i) {
+    // At this iteration, we replace node_args[i] with 
+    ORT_ENFORCE(node_args[i]->Name() != new_node_args[i]->Name());
+    ORT_ENFORCE(node_args[i]->Type() == new_node_args[i]->Type());
+
+    for (auto& node: nodes) {
+      for (auto& node_arg: node->MutableInputDefs()) {
+        if (node_arg->Name().compare(node_args[i]->Name()) != 0) {
+          continue;
+        }
+        node_arg = new_node_args[i];
+      }
+    }
+  }
+}
+
+std::string AddEventBeforeNode(
+  Graph& graph,
+  Node* node,
+  const std::string event_op_type,
+  const std::string event_op_name,
+  const std::string event_id_name) {
+  if (!node) {
+    // No event operator is be inserted, so we don't have its input event name.
+    return "";
+  }
+
+  // Inputs of "node" should be detached.
+  // "node" should consume outputs of the added event operator.
+  std::vector<NodeArg*> node_args = node->MutableInputDefs();
+
+  // Declare outputs of the added event operator.
+  std::vector<NodeArg*> new_node_args = CreateNewNodeArgs(graph, node_args);
+
+  // Convert Node* to std::vector<Node*>.
+  std::vector<Node*> nodes = {node};
+
+  // Replace node_args[i] with new_node_args[i] in nodes.
+  ReplaceNodeArgs(nodes, node_args, new_node_args); 
+
+  // Create node_arg for event ID.
+  auto event_node_arg = &CreateInt64NodeArg(graph, event_id_name);
+
+  // Create node which produces new_node_args from event ID and node_args.
+  auto name = graph.GenerateNodeName(event_op_name);
+  CreateBottleneckNode(graph,
+                       event_op_type,
+                       name,
+                       "",
+                       event_node_arg,
+                       node_args,
+                       new_node_args);
+
+  return event_node_arg->Name();
+}
+
+std::string AddEventAfterNode(
+  Graph& graph,
+  Node* node,
+  const std::string event_op_type,
+  const std::string event_op_name,
+  const std::string event_id_name) {
+  if (!node) {
+    // No event operator is be inserted, so we don't have its input event name.
+    return "";
+  }
```

Comment:
when will this be nullptr?

Reply:
Last stage doesn't have forward `Send` and backward `Recv`.

---
In reply to: [425509688](https://github.com/microsoft/onnxruntime/pull/3823#discussion_r425509688) [](ancestors = 425509688)

Reply:
Right, but in those cases, we don't call AddForwardWaitAfterRecv or AddBackwardWaitAfterRecv which use this function?

---
In reply to: [425996730](https://github.com/microsoft/onnxruntime/pull/3823#discussion_r425996730) [](ancestors = 425996730,425509688)

Reply:
You're right. It doesn't produce nullptr anymore.

---
In reply to: [426080509](https://github.com/microsoft/onnxruntime/pull/3823#discussion_r426080509) [](ancestors = 426080509,425996730,425509688)